File: ajoutproduit.js
let produitEnCours = null;

// 🔹 Ajouter un produit
document.getElementById("produitForm").addEventListener("submit", async function(event) {
    event.preventDefault();

    const produit = {
        libelle: document.getElementById("libelle").value,
        qualite: document.getElementById("qualite").value,
        prix: parseFloat(document.getElementById("prix").value),
        categorie: document.getElementById("categorie").value,
        stock: parseInt(document.getElementById("stock").value) || 0,
        imageProduit: document.getElementById("imageProduit").value,
        idVendeur: parseInt(document.getElementById("idVendeur").value)
    };

    try {
        const response = await fetch("http://127.0.0.1:8000/produits", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(produit)
        });

        if (!response.ok) {
            const error = await response.json();
            document.getElementById("message").textContent = "Erreur : " + error.detail;
            return;
        }

        const data = await response.json();
        document.getElementById("message").textContent = data.message;

        // Rafraîchir la liste des produits du vendeur
        chargerProduits(produit.idVendeur);

    } catch (err) {
        document.getElementById("message").textContent = "Erreur réseau : " + err;
    }
});

// 🔹 Charger les produits par vendeur
async function chargerProduits(idVendeur) {
    try {
        const response = await fetch(`http://127.0.0.1:8000/produits/vendeur/${idVendeur}`);
        const data = await response.json();

        const tbody = document.getElementById("produitsTable");
        tbody.innerHTML = "";

        data.produits.forEach(p => {
            const row = `<tr>
                <td>${p.idProduit}</td>
                <td>${p.libelle}</td>
                <td>${p.qualite || ""}</td>
                <td>${p.prix}</td>
                <td>${p.categorie || ""}</td>
                <td>${p.stock}</td>
                <td>${p.imageProduit ? `<img src="${p.imageProduit}" alt="image" width="50">` : ""}</td>
                <td>${p.idVendeur}</td>
                <td>
                    <button onclick="supprimerProduit(${p.idProduit}, ${p.idVendeur})">Supprimer</button>
                    <button onclick="ouvrirFormulaireModification(${p.idProduit}, ${p.idVendeur}, '${p.libelle}', '${p.qualite}', ${p.prix}, '${p.categorie}', ${p.stock}, '${p.imageProduit}')">Modifier</button>
                </td>
            </tr>`;
            tbody.innerHTML += row;
        });
    } catch (err) {
        document.getElementById("message").textContent = "Erreur lors du chargement des produits : " + err;
    }
}

// 🔹 Supprimer un produit
async function supprimerProduit(idProduit, idVendeur) {
    if (confirm("Voulez-vous vraiment supprimer ce produit ?")) {
        await fetch(`http://127.0.0.1:8000/produits/${idProduit}`, { method: "DELETE" });
        chargerProduits(idVendeur);
    }
}

// 🔹 Ouvrir le formulaire de modification avec les valeurs actuelles
function ouvrirFormulaireModification(idProduit, idVendeur, libelle, qualite, prix, categorie, stock, imageProduit) {
    produitEnCours = { idProduit, idVendeur };
    document.getElementById("editLibelle").value = libelle;
    document.getElementById("editQualite").value = qualite;
    document.getElementById("editPrix").value = prix;
    document.getElementById("editCategorie").value = categorie;
    document.getElementById("editStock").value = stock;
    document.getElementById("editImageProduit").value = imageProduit;
    document.getElementById("editForm").style.display = "block";
}

// 🔹 Valider la modification
async function validerModification() {
    const produit = {
        libelle: document.getElementById("editLibelle").value,
        qualite: document.getElementById("editQualite").value,
        prix: parseFloat(document.getElementById("editPrix").value),
        categorie: document.getElementById("editCategorie").value,
        stock: parseInt(document.getElementById("editStock").value),
        imageProduit: document.getElementById("editImageProduit").value,
        idVendeur: produitEnCours.idVendeur
    };

    await fetch(`http://127.0.0.1:8000/produits/${produitEnCours.idProduit}`, {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(produit)
    });

    document.getElementById("editForm").style.display = "none";
    chargerProduits(produitEnCours.idVendeur);
}

File: ajoutproduit.py
from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel
from datetime import date
import mysql.connector
from typing import Optional

app = FastAPI(title="API Produits - projet_ia")

#  Middleware CORS(au cas ou le front et le back n'utilisent pas le meme port)
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # en production : préciser le domaine du front
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

#  Fonction de connexion MySQL
def get_db_connection():
    return mysql.connector.connect(
        host="localhost",
        user="root",
        password="",
        database="projet_ia"
    )

#  Modèle de données
class Produit(BaseModel):
    libelle: str
    qualite: Optional[str] = None
    prix: float
    categorie: Optional[str] = None
    stock: Optional[int] = 0
    imageProduit: Optional[str] = None
    idVendeur: int

#  Ajouter un produit
@app.post("/produits", status_code=201)
def ajouter_produit(produit: Produit):
    if produit.prix <= 0:
        raise HTTPException(status_code=400, detail="Le prix doit être supérieur à 0")
    if produit.stock is not None and produit.stock < 0:
        raise HTTPException(status_code=400, detail="Le stock ne peut pas être négatif")

    sql = """
        INSERT INTO produit
        (libelle, qualite, prix, categorie, datePublication, stock, imageProduit, idVendeur)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
    """
    values = (
        produit.libelle,
        produit.qualite,
        produit.prix,
        produit.categorie,
        date.today(),
        produit.stock,
        produit.imageProduit,
        produit.idVendeur
    )

    db = get_db_connection()
    try:
        cursor = db.cursor()
        cursor.execute(sql, values)
        db.commit()
        cursor.close()
        db.close()
    except mysql.connector.Error as err:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Erreur lors de l'insertion : {err.msg}")

    return {"message": "Produit ajouté avec succès", "produit": produit}

#  Lister tous les produits
@app.get("/produits")
def lister_produits():
    db = get_db_connection()
    cursor = db.cursor(dictionary=True)
    cursor.execute("SELECT * FROM produit")
    produits = cursor.fetchall()
    cursor.close()
    db.close()
    return {"produits": produits}

#  Lister les produits par vendeur
@app.get("/produits/vendeur/{idVendeur}")
def produits_par_vendeur(idVendeur: int):
    db = get_db_connection()
    cursor = db.cursor(dictionary=True)
    cursor.execute("SELECT * FROM produit WHERE idVendeur = %s", (idVendeur,))
    produits = cursor.fetchall()
    cursor.close()
    db.close()
    return {"produits": produits}

#  Supprimer un produit
@app.delete("/produits/{idProduit}")
def supprimer_produit(idProduit: int):
    db = get_db_connection()
    cursor = db.cursor()
    try:
        cursor.execute("DELETE FROM produit WHERE idProduit = %s", (idProduit,))
        db.commit()
    except mysql.connector.Error as err:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Erreur lors de la suppression : {err.msg}")
    finally:
        cursor.close()
        db.close()
    return {"message": f"Produit {idProduit} supprimé avec succès"}

#  Modifier un produit
@app.put("/produits/{idProduit}")
def modifier_produit(idProduit: int, produit: Produit):
    db = get_db_connection()
    cursor = db.cursor()
    sql = """
        UPDATE produit
        SET libelle=%s, qualite=%s, prix=%s, categorie=%s, stock=%s, imageProduit=%s, idVendeur=%s
        WHERE idProduit=%s
    """
    values = (
        produit.libelle,
        produit.qualite,
        produit.prix,
        produit.categorie,
        produit.stock,
        produit.imageProduit,
        produit.idVendeur,
        idProduit
    )
    try:
        cursor.execute(sql, values)
        db.commit()
    except mysql.connector.Error as err:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Erreur lors de la modification : {err.msg}")
    finally:
        cursor.close()
        db.close()
    return {"message": f"Produit {idProduit} modifié avec succès", "produit": produit}
